server/interceptor/interceptor.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package interceptor defines gRPC interceptors for Trillian.
+package interceptor
+
+import (
+	"github.com/google/trillian"
+	"github.com/google/trillian/errors"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/trees"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// TreeInterceptor ensures that all requests pertaining a specific tree are of correct type and
+// respect the current tree state (frozen, deleted, etc).
+type TreeInterceptor struct {
+	Admin storage.AdminStorage
+}
+
+// UnaryInterceptor executes the TreeInterceptor logic for unary RPCs.
+func (i *TreeInterceptor) UnaryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	rpcInfo, err := getRPCInfo(req)
+	if err != nil {
+		return nil, err
+	}
+
+	if rpcInfo.doesNotHaveTree {
+		return handler(ctx, req)
+	}
+
+	tree, err := trees.GetTree(ctx, i.Admin, rpcInfo.treeID, rpcInfo.opts)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx = trees.NewContext(ctx, tree)
+	return handler(ctx, req)
+}
+
+// rpcInfo contains information about an RPC, as extract from its request message.
+type rpcInfo struct {
+	// doesNotHaveTree states whether the RPC is tied to a specific tree.
+	// Examples of RPCs without trees are CreateTree (no tree exists yet) and ListTrees
+	// (zero to many trees returned).
+	doesNotHaveTree bool
+	// treeID is the tree ID tied to this RPC, if any.
+	treeID int64
+	// opts is the trees.GetOpts appropriate to this RPC (TreeType, readonly vs readwrite, etc).
+	opts trees.GetOpts
+}
+
+// getRPCInfo returns the rpcInfo for the given request, or an error if the request is not mapped.
```

Comment:
I'm nervous about having metadata about the RPCs buried away separately from the RPC definitions -- it will be harder to keep in sync, and harder to debug when things do get out of sync.

Random ideas:
 - We could partly help by documenting around it (e.g. add reminders in *.proto)
 - Maybe investigate whether there's some way of tagging the RPCs with metadata information.
 - Impose (and document) some conventions for RPCs and use that to autogenerate.

----
(Random musings continue)

For the last of those, one thought would be to assume:
 - any request whose name starts with `Get..` is read-only (writable o/w)
 - if a request has a `GetMapId` method, that provides the `treeID` and `opts.TreeType` is MAP
 - if a request has a `GetLogId` method, that provides the `treeID` and `opts.TreeType` is LOG
 - if a request has a `GetTreeId` method, that provides the `treeID` and `opts.TreeType` is clear
 - otherwise, the request `doesNotHaveTree`

To fulfill those assumptions, I think the only needed change to the RPC definitions would be to duplicate `UpdateTreeRequest.Tree.TreeID` as a separate `UpdateTreeRequest.TreeID` field; then you could have:

```go
func getRPCInfo(req interface{}) (*rpcInfo, error) {
	// Get.. methods on a specific tree are assumed to be read-only.
	reqType := reflect.TypeOf(req)
	isGetReq := strings.HasPrefix(reqType.Name(), "Get")

	if i, ok := req.(interface { GetMapId() int64 }); ok {
		// The ..Request message has a MapId field, which provides the treeID (and
		// means this must be map-only).
		return &rpcInfo{
			treeID: i.GetMapId(),
			opts: trees.GetOpts{
				TreeType: trillian.TreeType_MAP,
				Readonly: isGetReq,
			},
		}, nil
	}

	if i, ok := req.(interface { GetLogId() int64 }); ok {
		// The ..Request message has a MapId field, which provides the treeID (and
		// means this must be log-only).
		return &rpcInfo{
			treeID: i.GetLogId(),
			opts: trees.GetOpts{
				TreeType: trillian.TreeType_LOG,
				Readonly: isGetReq,
			},
		}, nil
	}

	if i, ok := req.(interface { GetTreeId() int64 }); ok {
		// The ..Request message has a TreeId field.
		return &rpcInfo{
			treeID: i.GetTreeId(),
			opts: trees.GetOpts{
				Readonly: isGetReq,
			},
		}, nil
	}

	return &rpcInfo{doesNotHaveTree: true}, nil
}
```

OTOH, this may well be too obscure / action-at-a-distance and so would hinder rather than help with the keeping-in-sync problem.